packages/frontend/src/hooks/useDynamicData.ts

```diff
@@ -45,57 +57,65 @@ function computeArguments(
  * @param schema - the field that needs the dynamic data
  */
 function useDynamicData(stepId: string | undefined, schema: IField) {
-  const lastComputedVariables = React.useRef({})
-  const [getDynamicData, { called, data, loading, refetch }] =
-    useLazyQuery(GET_DYNAMIC_DATA)
-  const { getValues } = useFormContext()
-
-  /**
-   * Return `null` when even a field is missing value.
-   *
-   * This must return the same reference if no computed variable is changed.
-   * Otherwise, it causes redundant network request!
-   */
-  const computedVariables = React.useMemo(() => {
-    if (schema.type === 'dropdown' && schema.source) {
-      try {
-        const variables = computeArguments(schema.source.arguments, getValues)
-
-        // if computed variables are the same, return the last computed variables.
-        if (isEqual(variables, lastComputedVariables.current)) {
-          return lastComputedVariables.current
-        }
-
-        lastComputedVariables.current = variables
-
-        return variables
-      } catch (err) {
-        return null
+  const { getValues, watch } = useFormContext()
+  const { nonFormFieldArgs, watchedFormFields } = useMemo(() => {
+    if (schema.type !== 'dropdown' || !schema.source) {
+      return {
+        nonFormFieldArgs: {},
+        watchedFormFields: new Map(),
       }
     }
 
-    return null
-    /**
-     * `formValues` is to trigger recomputation when form is updated.
-     * `getValues` is for convenience as it supports paths for fields like `getValues('foo.bar.baz')`.
-     */
-  }, [schema, getValues])
-
-  React.useEffect(() => {
-    if (
-      schema.type === 'dropdown' &&
-      stepId &&
-      schema.source &&
-      computedVariables
-    ) {
-      getDynamicData({
-        variables: {
-          stepId,
-          ...computedVariables,
-        },
-      })
+    return processArguments(schema.source.arguments)
+  }, [schema])
+  const watchedFormFieldValues = useMemo(
+    () => getWatchedFormFieldValues(watchedFormFields, getValues()),
+    [
+      watchedFormFields,
+      // We depend on getValues instead of its result, because we don't want to
+      // re-query each time a form value changes. Instead, we use watch (below)
+      // on the fields we're interested in.
+      getValues,
+    ],
+  )
+
+  const shouldSkipQuery =
+    !stepId || schema.type !== 'dropdown' || !schema.source
+  const { called, data, loading, refetch } = useQuery(GET_DYNAMIC_DATA, {
+    variables: {
+      stepId,
+      ...nonFormFieldArgs,
+      ...watchedFormFieldValues,
+    },
+    skip: shouldSkipQuery,
+    notifyOnNetworkStatusChange: true,
+  })
+
+  // Refetch if any of our watched form fields changes.
```

Comment:
noob question but when does the watched form fields change ah, a bit unclear how this works?